libs/features/mas/web-components/src/aem.js

```diff
@@ -1,63 +1,211 @@
-const accessToken = localStorage.getItem('masAccessToken');
-
-const headers = {
-    Authorization: `Bearer ${accessToken}`,
-    pragma: 'no-cache',
-    'cache-control': 'no-cache',
-};
-
-/**
- * Search for content fragments
- * @param {Object} params - The search options
- * @param {string} [params.path] - The path to search in
- * @param {string} [params.query] - The search query
- * @returns {Promise<Array>} - A promise that resolves to an array of search results
- */
-async function fragmentSearch({ path, query }) {
-    const filter = {};
-    if (path) {
-        filter.path = path;
-    }
-    if (query) {
-        filter.fullText = {
-            text: encodeURIComponent(query),
-            queryMode: 'EXACT_WORDS',
+import { wait } from './utils.js';
+
+class AEM {
+    #author;
+    constructor(bucket) {
+        this.#author = /^author-/.test(bucket);
+        const baseUrl = `https://${bucket}.adobeaemcloud.com`;
+        const sitesUrl = `${baseUrl}/adobe/sites`;
+        this.cfFragmentsUrl = `${sitesUrl}/cf/fragments`;
+        this.cfSearchUrl = `${this.cfFragmentsUrl}/search`;
+        this.cfPublishUrl = `${this.cfFragmentsUrl}/publish`;
+        this.wcmcommandUrl = `${baseUrl}/bin/wcmcommand`;
+        this.csrfTokenUrl = `${baseUrl}/libs/granite/csrf/token.json`;
+
+        this.headers = {
+            // IMS users might not not have all the permissions, token in the sessionStorage is a temporary workaround
+            Authorization: `Bearer ${sessionStorage.getItem('masAccessToken') ?? window.adobeid?.authorize?.()}`,
+            pragma: 'no-cache',
+            'cache-control': 'no-cache',
         };
     }
-    const searchParams = new URLSearchParams({
-        query: JSON.stringify({ filter }),
-    }).toString();
-    return fetch(`${this.cfSearchUrl}?${searchParams}`, {
-        headers,
-    })
-        .then((res) => res.json())
-        .then(({ items }) => items);
-}
 
-async function getCfByPath(path) {
-    return fetch(`${this.cfFragmentsUrl}?path=${path}`, {
-        headers,
-    })
-        .then((res) => res.json())
-        .then(({ items: [item] }) => item);
-}
+    async getCsrfToken() {
+        const { token } = await fetch(this.csrfTokenUrl, {
```

Comment:
Should we include error handling in some or every `fetch` to manage network errors or unexpected responses?

Reply:
Ideally yes. This is needed in MAS Studio only and actually a temporary solution, and needs to be removed once the AEM eng fixes a bug I found in the CF cloning.
Also, error handling is implemented on the consumer of aem.js since the only choice we have in aem.js is to throw an Error.